File: enode/src/main/java/org/enodeframework/domain/impl/DefaultRepository.java
package org.enodeframework.domain.impl;

import org.enodeframework.common.utils.Assert;
import org.enodeframework.domain.AggregateRoot;
import org.enodeframework.domain.MemoryCache;
import org.enodeframework.domain.Repository;

import java.util.concurrent.CompletableFuture;

/**
 * @author anruence@gmail.com
 */
public class DefaultRepository implements Repository {

    private final MemoryCache memoryCache;

    public DefaultRepository(MemoryCache memoryCache) {
        this.memoryCache = memoryCache;
    }

    @Override
    public <T extends AggregateRoot> CompletableFuture<T> getAsync(Class<T> aggregateRootType, Object aggregateRootId) {
        Assert.nonNull(aggregateRootType, "aggregateRootType");
        Assert.nonNull(aggregateRootId, "aggregateRootId");
        CompletableFuture<T> future = memoryCache.getAsync(aggregateRootId, aggregateRootType);
        return future.thenCompose(aggregateRoot -> {
            if (aggregateRoot == null) {
                return memoryCache.refreshAggregateFromEventStoreAsync(aggregateRootType, aggregateRootId.toString());
            }
            return CompletableFuture.completedFuture(aggregateRoot);
        });
    }

    @Override
    public CompletableFuture<AggregateRoot> getAsync(Object aggregateRootId) {
        return getAsync(AggregateRoot.class, aggregateRootId);
    }

    @Override
    public <T extends AggregateRoot> void refreshAggregate(T aggregateRoot) {
        memoryCache.refreshAggregate(aggregateRoot);
    }
}

File: enode/src/main/java/org/enodeframework/domain/impl/DefaultRepository.kt
package org.enodeframework.domain.impl

import org.enodeframework.common.utils.Assert
import org.enodeframework.domain.AggregateRoot
import org.enodeframework.domain.MemoryCache
import org.enodeframework.domain.Repository
import java.util.concurrent.CompletableFuture

/**
 * @author anruence@gmail.com
 */
class DefaultRepository(private val memoryCache: MemoryCache) : Repository {
    override fun <T : AggregateRoot?> getAsync(
        aggregateRootType: Class<T>,
        aggregateRootId: String
    ): CompletableFuture<T> {
        Assert.nonNull(aggregateRootType, "aggregateRootType")
        Assert.nonNull(aggregateRootId, "aggregateRootId")
        var result = CompletableFuture<T>()
        memoryCache.getAsync(aggregateRootId, aggregateRootType).whenComplete { aggregateRoot: T, _ ->
            if (aggregateRoot != null) {
                result.complete(aggregateRoot)
            } else {
                result = memoryCache.refreshAggregateFromEventStoreAsync(
                    aggregateRootType,
                    aggregateRootId
                )
            }
        }
        return result
    }

    override fun <T : AggregateRoot?> refreshAggregate(aggregateRoot: T) {
        memoryCache.refreshAggregate(aggregateRoot!!)
    }
}
